### PR TITLE
Rabbitmq eventbus implementation

### PR DESCRIPTION
**Pull request by trantienduchn**

### CODE REVIEW

File: backends-common/rabbitmq/src/test/java/org/apache/james/backend/rabbitmq/RabbitMQFixture.java [hunk 1]
```diff
@@ -37,6 +38,7 @@
     public static final String DIRECT = "direct";
     public static final boolean EXCLUSIVE = true;
     public static final boolean AUTO_DELETE = true;
+    public static final ImmutableMap<String, Object> NO_ARGUMENTS = ImmutableMap.of();
```
Comment (chibenwa) [inline backends-common/rabbitmq/src/test/java/org/apache/james/backend/rabbitmq/RabbitMQFixture.java]: Most of the contant of this class are not related to tests and deserves to be moved to main `Constant.java`

File: mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/RoutingKeyConverter.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    static String toRoutingKey(RegistrationKey key) {
+        return key.getClass().getName() + SEPARATOR + key.asString();
+    }
+
+    @VisibleForTesting
+    static RoutingKeyConverter forFactories(RegistrationKey.Factory... factories) {
+        return new RoutingKeyConverter(ImmutableSet.copyOf(factories));
+    }
+
+    private static final String SEPARATOR = ":";
```
Comment (chibenwa) [inline mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/RoutingKeyConverter.java]: static constants should go before static methods

**Comment by chibenwa** [comment]
```
[a833f2a8b8a7c76427b19f1297354a4a0f6a9db0] [ERROR]   SeveralRabbitMQEventBusTest.groupListenersShouldBeExecutedOnceInAControlledEnvironment:146 
[a833f2a8b8a7c76427b19f1297354a4a0f6a9db0] Wanted but not invoked:
[a833f2a8b8a7c76427b19f1297354a4a0f6a9db0] mailboxListener.event(<any>);
[a833f2a8b8a7c76427b19f1297354a4a0f6a9db0] -> at org.apache.james.mailbox.events.SeveralRabbitMQEventBusTest.groupListenersShouldBeExecutedOnceInAControlledEnvironment(SeveralRabbitMQEventBusTest.java:146)
[a833f2a8b8a7c76427b19f1297354a4a0f6a9db0] Actually, there were zero interactions with this mock.
```

**Comment by trantienduchn** [comment]
when there is no listener registered, then events dipatched are all ignored to handle. Is it expected? Should we store events temporally in queue and waits util there are at least one registration for those events?

**Comment by chibenwa** [comment]
> when there is no listener registered, then events dipatched are all ignored to handle. Is it expected?

Yes this is expected.

> Should we store events temporally in queue and waits util there are at least one registration for those events?

If nobody is registered then nobody is interested.

File: mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/RegistrationBinder.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        this.registrationQueue = registrationQueue;
+    }
+
+    Mono<Void> bind(RegistrationKey key) {
+        return sender.bind(bindingSpecification(key))
+            .then();
+    }
+
+    Mono<Void> unbind(RegistrationKey key) {
+        return sender.bind(bindingSpecification(key))
```
Comment (aduprat) [inline mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/RegistrationBinder.java]: unbind?

File: mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/RoutingKeyConverter.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+import com.google.common.base.Preconditions;
+import com.google.common.base.Splitter;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+
+class RoutingKeyConverter {
+    private static final String SEPARATOR = ":";
+
+    static String toRoutingKey(RegistrationKey key) {
+        return key.getClass().getName() + SEPARATOR + key.asString();
```
Comment (blackheaven) [inline mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/RoutingKeyConverter.java]: I'm a bit annoyed by that `String` manipulation, I think creating a dedicated class will be better

Comment (chibenwa) [reply mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/RoutingKeyConverter.java]: Agree, RoutingKey strong typing could be nice!

File: mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/GroupRegistration.java [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        try {
+            mailboxListener.event(event);
+        } catch (Exception e) {
+            LOGGER.error("Exception happens when handling event of user {}", event.getUser().asString(), e);
+        }
+    }
+
+    @Override
+    public void unregister() {
+        if (receiverSubscriber != null && !receiverSubscriber.isDisposed()) {
```
Comment (blackheaven) [inline mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/GroupRegistration.java]: Can you make `receiverSubscriber` an `Optional`?

File: mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/RoutingKeyConverter.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+    private final Set<RegistrationKey.Factory> factories;
+
+    @Inject
+    RoutingKeyConverter(Set<RegistrationKey.Factory> factories) {
+        this.factories = factories;
+    }
+
+    RegistrationKey toRegistrationKey(String routingKey) {
+        return toRegistrationKey(Splitter.on(SEPARATOR).splitToList(routingKey));
```
Comment (aduprat) [inline mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/RoutingKeyConverter.java]: Maybe using the `on(Pattern)` method help you by splitting only the first occurrence of `String:`
then you will not need to join in order to recreate the value.

File: mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/KeyRegistrationHandler.java [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+        this.sender = sender;
+        this.routingKeyConverter = routingKeyConverter;
+        this.mailboxListenerRegistry = new MailboxListenerRegistry();
+        this.receiver = RabbitFlux.createReceiver(new ReceiverOptions().connectionMono(connectionMono));
+        this.registrationQueue = new RegistrationQueueName();
+        this.registrationBinder = new RegistrationBinder(sender, registrationQueue);
+    }
+
+    void start() {
+        sender.declareQueue(QueueSpecification.queue()
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/KeyRegistrationHandler.java]: Non durable  means a greater inconsistency window during the start process (as events received between queue binding & queue consumption will be lost)

The benefits of being non-durable is very certainly performance.

This seems to make sens for this feature - but for the sake of completness, is this trade-off documented?

Also, let's avoid implicit rabbitMQ config, and keep these parameter explicits as they are far from easy to choose. Furthermore, explicitly having this might defend us against driver changes.

[the remark this answers is not in the document]
Comment (trantienduchn) [reply mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/KeyRegistrationHandler.java]: then you have to specify empty string name to the queue. 

I thought in the meeting about presenting rabbitmq eventbus yesterday, you expected it was non durable. its queue name is randomly, when you stop eventbus, and start again, I will create another queue for key registration with different name, you could not re use the old queue and events in that queue.
The second point is, you stop an eventbus, if I remember well, key registration will be wiped out, then all events will be lost anyway

**Comment by chibenwa** [comment]
Was :green_apple: 

test this please

**Comment by openpaas-james** [comment]
Build triggered. sha1 is merged.

**Comment by openpaas-james** [comment]
Build started sha1 is merged.

**Comment by chibenwa** [comment]
Handled in regular PR of https://github.com/linagora/james-project/milestone/14
